### PR TITLE
Add a basic ghost package

### PR DESCRIPTION
**Pull request by yuwenma**
This package is originated from helm chart bitnami/ghost, which uses MariaDB.
By default, the ghost helmchart does not have the ghost, nor does it enable ingress, volumePermissions, networkPolicy, prometheus.

When rendering the ghost helm chart, we include the ghost host (`Deployment` and `PersistentVolumeClaim`), as well as the other services (`NetworkPolicy` `Ingress` `PrometheusRule`). 

Follow up work is tracked in https://github.com/GoogleContainerTools/kpt/issues/3379

### CODE REVIEW

File: package-examples/ghost/Kptfile
Comment (droot) [inline package-examples/ghost/Kptfile:11]: Was curious we aren't using `set-namespace:v0.4.1` ?

File: package-examples/ghost/ghost/Kptfile
Comment (droot) [inline package-examples/ghost/ghost/Kptfile:4]: ghost pkg inside ghost confused me a bit :) Will think of an alternative as I go further in the PR.

File: package-examples/ghost/ghost/deployment-example-ghost.yaml
Comment (droot) [inline package-examples/ghost/ghost/deployment-example-ghost.yaml:21]: This is interesting for `set-namespace`'s function perspective.

File: package-examples/ghost/ghost/secret-example-ghost.yaml
Comment (droot) [inline package-examples/ghost/ghost/secret-example-ghost.yaml:8]: We should not have this secret in this package. 

The Good thing is when user will deploy this package, the deployment that needs this secret will be blocked until the secret is available.

So for now secret can be provisioned out-of-band manually. And in later part of our CUJ, we can add a secret manager to manage the secret. WDYT ?

**Review by droot — COMMENTED** [review]
Some quick comments. Feel free to merge it and address it as a follow up PR.

File: package-examples/ghost/ghost/deployment-example-ghost.yaml
Comment (bgrant0607) [inline package-examples/ghost/ghost/deployment-example-ghost.yaml:28]: I wouldn't prefix the container name. There's no real point to it.

File: package-examples/ghost/ghost/deployment-example-ghost.yaml
Comment (bgrant0607) [inline package-examples/ghost/ghost/deployment-example-ghost.yaml:4]: Variants are being uniquified by name prefix as well as by namespace name?

File: package-examples/ghost/ghost/secret-example-ghost.yaml
Comment (bgrant0607) [inline package-examples/ghost/ghost/secret-example-ghost.yaml:1]: This is ok for the initial version, but we don't want actual Secret resources in packages, especially not in git. We'll want a configuration that specifies how to generate or access the Secret instead. See #3125

File: package-examples/ghost/Kptfile
Comment (yuwenma) [reply package-examples/ghost/Kptfile:11]: `v0.4.1` does not accept duplicate GVKNN which means using `v0.4.1` users need to specify the set-namespace in the children package rather than using it in the parent. 
FYI, I think that is an unnecessary requirement. So the latest set-namespace (unreleased) removes this requirement.

File: package-examples/ghost/ghost/deployment-example-ghost.yaml
Comment (yuwenma) [reply package-examples/ghost/ghost/deployment-example-ghost.yaml:21]: Yes, so as the namespaceSelector for podAffinityTerm.

File: package-examples/ghost/mariadb/configmap-example-mariadb.yaml
Comment (yuwenma) [reply package-examples/ghost/mariadb/configmap-example-mariadb.yaml:8]: Yes! This package gives the example for issues  
https://github.com/GoogleContainerTools/kpt/issues/3118
https://github.com/GoogleContainerTools/kpt/issues/3119
https://github.com/GoogleContainerTools/kpt/issues/3210 (Augeas part)

File: package-examples/ghost/ghost/deployment-example-ghost.yaml
Comment (bgrant0607) [reply package-examples/ghost/ghost/deployment-example-ghost.yaml:13]: https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments

File: package-examples/ghost/package-context.yaml
Comment (yuwenma) [reply package-examples/ghost/package-context.yaml:4]: Both `kpt pkg init` and `kpt pkg get --for-deployment` generate the package-context.yaml. `kpt pkg get` won't.

This package-contex.yaml is generated by `kpt pkg init`, which is how we suggest users to create blueprint from scratch in kpt CLI https://kpt.dev/guides/namespace-provisioning-cli?id=initialize-a-package. @droot shall we change kpt on `kpt pkg init` to not creating `package-context.yaml`? But then it will break the local testing. Or we can add one more step to delete this files manually.

File: package-examples/ghost/mariadb/configmap-example-mariadb.yaml
Comment (bgrant0607) [reply package-examples/ghost/mariadb/configmap-example-mariadb.yaml:8]: Yep.

If this is INI format, I'd change the file suffix to .ini and change the corresponding config mount in the StatefulSet.

File: package-examples/ghost/mariadb/package-context.yaml
Comment (bgrant0607) [reply package-examples/ghost/mariadb/package-context.yaml:8]: A user in slack has encountered similar issues using Config Connector subpackages (e.g., CloudSQL).

File: package-examples/ghost/ghost/setlabels.yaml
Comment (yuwenma) [reply package-examples/ghost/ghost/setlabels.yaml:7]: 😞 no spec. I think all custom functionConfig type do not follow the `spec` `status` rule. The argument is that those resources won't be deployed to the cluster so it is not a strict spec/status. Skipping spec saves some manual editing. 

I'm more happy to add the `spec` to `functionConfig`, because it does help answering questions like "what fields & resources a user could be interested in modifying"  https://kubernetes.slack.com/archives/C0155NSPJSZ/p1657561360040189

File: package-examples/ghost/ghost/secret-example-ghost.yaml
Comment (droot) [reply package-examples/ghost/ghost/secret-example-ghost.yaml:1]: We discussed and decided to treat secret as an external resources (not managed by our stack for now). So secrets will not be part of the package.

File: package-examples/ghost/ghost/secret-example-ghost.yaml
Comment (droot) [reply package-examples/ghost/ghost/secret-example-ghost.yaml:8]: This is discussed offline...treating secret as an external resources.

File: package-examples/ghost/ghost/setlabels.yaml
Comment (droot) [reply package-examples/ghost/ghost/setlabels.yaml:9]: Yeah, I was wondering if the current package wires using the ApplyReplacements or Starlark.

File: package-examples/ghost/package-context.yaml
Comment (droot) [reply package-examples/ghost/package-context.yaml:4]: @bgrant0607 your understanding is correct and current behavior is inline with that.

@yuwenma . No, we don't have to change the `kpt pkg init` behavior. 

Listing the scenario that we need to confirm:

kpt pkg get <ghost-super-pkg> app-1 --for-deployment

1. the app-1/package-context.yaml will contain `data.name: app-1`
2. app-1/ghost/package-context.yaml  ??
3. app-2/mariadb/package-context.yaml ??

And when you run `kpt fn render app-1`, the set-namespace will use the package-context at the top-level, so everything should be fine, but we need to confirm this.

**Review by droot — COMMENTED** [review]
We discussed this offline and I captured the notes here.

File: package-examples/ghost/ghost/Kptfile
Comment (yuwenma) [reply package-examples/ghost/ghost/Kptfile:4]: chose ghost-app. Updated.

File: package-examples/ghost/package-context.yaml
Comment (yuwenma) [reply package-examples/ghost/package-context.yaml:4]: I added the full steps in README.yaml and approved this works well:

`kpt pkg get --for-deployment` will create package-context.yaml under each sub directory as well as the parent directory, with the directory name as the `data.name`, which are `ghost-app`, `mariadb` `root-dir`. Running set-namespace on root updates all namespaces to the value in `root-dir` package context `data.name`. Then, `kpt live apply root-dir` deploys all resources to the cluster successfully

**Review by yuwenma — COMMENTED** [review]
Package updated, please take another look @bgrant0607 @droot 
Some key changes:
- Removed secrets and the references (env in pod template, checksum/secret annotation, livenessProbe).
- Added CRD for PrometheusRule under `crds`, this should be the only CRD needs to install.
- Dropped name prefix `example-` and updated the file names.
- Dropped the `app.kubernetes.io/instance` label, I leave a question about the kpt equivalence.
- Updated README to guide users get, hydrate and deploy the package.

File: package-examples/ghost/ghost/setlabels.yaml
Comment (bgrant0607) [reply package-examples/ghost/ghost/setlabels.yaml]: On the kpt side, the deployment package name is similar to the release name.

File: package-examples/ghost/ghost/Kptfile
Comment (bgrant0607) [reply package-examples/ghost/ghost/Kptfile]: I've been thinking we need to provide the full package path of any subpackage as a string, such as toppkg-middlepkg-leafpkg. Maybe that's what should populate the subpackage package-context. Or maybe we should add an additional field to it containing the full name.

For example, if the top-level package name were "ghost" and this subpackage's name were "app", then then full name would be "ghost-app".

File: package-examples/ghost/ghost/deployment-example-ghost.yaml
Comment (bgrant0607) [reply package-examples/ghost/ghost/deployment-example-ghost.yaml]: I prefer the way kustomize does this, by generating a new ConfigMap or Secret with a hash appended to the name. If the ConfigMap or Secret is modified in place, then rollout and rollback can't be controlled predictably.

That said, this really should be done server-side at this point.

**Comment by bgrant0607** [comment]
Thanks. LGTM if the application works.

File: package-examples/ghost/README.md
Comment (droot) [inline package-examples/ghost/README.md:12]: why can't we add namespace to the package itself ? 

If you used helm chart to get the manifests, you can use `--create-namespace` arg to add it.

That will make the workflow truly declarative.

**Review by droot — APPROVED** [review]
I have a minor comment, but it is good to go.

File: package-examples/ghost/README.md
Comment (bgrant0607) [reply package-examples/ghost/README.md:12]: The ability to create a namespace would imply that the app team had cluster admin privilege. It also would bring in the need for the flexibility of the namespace provisioning scenario we've been working on: resource quota, limit range, network policy, etc.

Comment (bgrant0607) [reply package-examples/ghost/README.md:12]: We should have an example namespace blueprint for this, which could be used instead of kubectl create namespace.

Comment (droot) [reply package-examples/ghost/README.md:12]: > The ability to create a namespace would imply that the app team had cluster admin privilege. It also would bring in the need for the flexibility of the namespace provisioning scenario we've been working on: resource quota, limit range, network policy, etc.

Aah.. completely missed that. Good pt.

**Pull request merged**
